### ISSUE TITLE
m68k support

### ISSUE THREAD

**Issue by mattgodbolt** [mattgodbolt/compiler-explorer#271]
George Nakos runs http://brownbot.hopto.org/ which has m68k support. I asked him how he did it and he gamely had a crack at a patch for CE. He had to bail as it got a little warty but here's his notes:

```
We build binutils using:

../binutils-2.27/configure --disable-multilib --disable-nls --enable-lto --prefix=/usr --target=m68k-ataribrown-elf

So, a plain binutils 2.27, nothing fancy at all. We chose the "ataribrown" name to differentiate from other versions of cross-gcc for ataris floating about. I guess you could use "unknown" so it's a bit more generic, after all there's no patching on this level at all so it should apply for all m68k based machines.

Next, we build gcc itself using:

   ../gcc-6.2.0/configure \
       --target=m68k-ataribrown-elf \
       --disable-nls \
       --enable-languages=c,c++ \
       --enable-lto \
       --prefix=/usr \
       --disable-libssp \
       --enable-softfloat \
       --disable-libstdcxx-pch \
       --disable-clocale \
       --disable-libstdcxx-threads \
       --disable-libstdcxx-filesystem-ts \
       --disable-libquadmath \
       --enable-cxx-flags='-O2 -fomit-frame-pointer -fno-threadsafe-statics -fno-exceptions -fno-rtti -fleading-underscore' \
       CFLAGS_FOR_TARGET="-O2 -fomit-frame-pointer -fno-threadsafe-statics -fleading-underscore" \
       CXXFLAGS_FOR_TARGET="-O2 -fomit-frame-pointer -fno-threadsafe-statics -fno-exceptions -fno-rtti -fleading-underscore" \
       LDFLAGS_FOR_TARGET="--emit-relocs -Ttext=0"

Most of the switches deal with switching off stuff that made compilation fail (exceptions for example, especially when building the libs).

From what I understand you don't require building any libs, right? If yes then this is pretty much it if I'm not mistaken. I've tried building it in linux mint 18 (which is ubuntu 16.06 or 16.04 - can't remember) so it should build cleanly.
```

**Comment by mattgodbolt** [comment]
cf #82

**Comment by ggnkua** [comment]
If anyone has any questions regarding the above, I'll be glad to help!

**Comment by ggnkua** [comment]
Ok, so I decided to overcome my fears and man up on this :)

First of all, am I correct in thinking that I have to add the compiler(s) to the live site repo? (https://github.com/mattgodbolt/compiler-explorer-image)

If that's the case then I think it's very easy to add my build script to the collection of scripts there. Sadly I don't provide binary distributions of the cross compilers but instead encourage people to build their own. However the build script I maintain over at https://bitbucket.org/ggnkua/bigbrownbuild is pretty much automated, just run it and it builds binutils, compiler and a library specific to Atari ST and compatibles.

So here are a couple of questions:

1. Would you agree that I have to add the compiler first at compiler-explorer-image repo and then at compiler-explorer?
1. I ran the install_compilers.sh script and noticed the folder structure at /opt/compiler-explorer. How would you suggest this line of compilers be named? Would something like gcc-x-x-x-68k be ok?
1. Would it be ok to build the compiler(s) from scratch during the update-compilers execution? I noticed that the scripts more or less fetch prebuilt binaries from various sources and place them at the proper folders. I have to admit that I still don't know much about how docker operates but this is my best assumption.

**Comment by RabsRincon** [comment]
Hi! Sorry for taking a bit to answer, and thanks for your interest in improving Compiler Explorer :)

You're right in most of the point you raise, but let me clarify some/give my input, and hopefully you find it helpful. If you need more help, please don't hesitate in contacting us again.

You're right in that the compilers have to be added both here, and in the sister repo.

For 2) may I suggest gcc68k-x-y-z? I think it's visually easier to identify those, but you're pretty much free to choose the scheme you find more suitable

As for building the compilers, this is a trickier subject. We build the necessary compilers with the help of the python files at the root of the sister repostory (and later store them in s3), so you can take a look at those too. @mattgodbolt will probably chime in after the weekend, and he'll probably be more helpful than I am in regards to this (Sorry!)

Let me know if I can help you with anything else

:)

**Comment by mattgodbolt** [comment]
@ggnkua: It's been nearly two years and I never replied: I'm so sorry!

1. Yes; ideally. Having a builder that runs in one of the `docker` containers and generates a tarball is the ideal case; and then writing the lines to unpack that tarball in the `update_compilers.sh` scripts
2. /opt/compiler-explorer/gcc-x-x-x-68k` would work yes. Probably better to be `/opt/compiler-explorer/68x/gcc-x-x-x` for tidyness sake though.
3. No; unfortunately this is not OK: the "update_compilers" is not even run on a node that can do much building. The building must happen in the docker containers.

That said: if you can take one of the existing build scripts (e.g.  something like https://github.com/mattgodbolt/compiler-explorer-image/blob/master/gcc/build/build-avr.sh ) and get it to work as a one-off outside of docker, we could handle the docker part for you (docker just gives us a reproducible environment so that years down the line we have a hope of recovering old compilers if we ever need to).

Hope that gives some pointers, and sorry it has been so very long!

**Comment by ggnkua** [comment]
Hi @mattgodbolt, thanks for your reply. You might want to check out the relevant issue I've opened in compiler-explorer image (https://github.com/mattgodbolt/compiler-explorer-image/issues/68 - very apt issue number btw!) where I've done some work setting it all up in a container for you.

I think we could move the discussion there for now and get back to this issue later!

**Comment by dforgeas** [comment]
The new address for brownbot is this by the way: http://brownbot.mooo.com/

**Comment by windenntw** [comment]
FYI m68k has now landed in LLVM: https://github.com/llvm/llvm-project/tree/main/llvm/lib/Target/M68k

**Comment by mattgodbolt** [comment]
@windelbouwman we build llvm trunk every night, so in theory one can pass the relevant params to get m68k, though I tried the obvious https://godbolt.org/z/aPq7WEnb9 and it didn't work. Perhaps we need to enable it in our build [here](https://github.com/compiler-explorer/clang-builder/blob/main/build/build.sh#L95) ?

**Comment by windenntw** [comment]
Hi @mattgodbolt 

TLDR: Yea, you'd need to enable m68k there.

Min-Yih mentions in https://groups.google.com/g/llvm-dev/c/KH5jlTH_is4 

I’m also working pretty hard recently on upstreaming Motorola M68k backend to LLVM [1], and the works are wrapping up so I might can give some advices
...

Usually a new target will enter LLVM as an experimental target first (it will not be build by default unless designated by the `LLVM_EXPERIMENTAL_TARGETS_TO_BUILD` cmake variable) before becoming an official one.
...

**Comment by DavidSpickett** [comment]
> TLDR: Yea, you'd need to enable m68k there.

https://github.com/compiler-explorer/clang-builder/pull/55

**Comment by dkm** [comment]
As m68k is now part of our cross targets for both gcc and llvm, we can close this :)
